### DISCUSSION TITLE
Choosing a suitable output configuration..

### DISCUSSION THREAD

**Discussion by pietervandermeer** [tomaka/cpal#1316] [Q&A] [answered]
I've used Cpal for a while now and imho it excels at abstracting away all the details from the underlying audio API's. Mostly it picks the right default configuration for you, which is good. But sometimes it doesn't. For instance, what happens when you have multiple sound cards with a wide variety of sample rates? Another thing is sample format.. Sometimes it returns I16, when the underlying API definitely supports F32.. I'd like to make explicit choices here, as I'm sure would many others. 

The enumerate example explains how to iterate through all possible configurations. But to actually match and choose one that's an entirely different thing. cpal::SupportedStreamConfigRange consists of private members, so how are we supposed to match a suitable configuration? More examples would definitely be welcome! Thanks in advance.

**Comment by pietervandermeer** [comment, marked as answer]
Ok, I found out how to pick a specific configuration by comparing. `SupportedStreamConfigRange` has member functions to read out the specs. `pub fn channels(&self) -> ChannelCount` for instance. 

There's something weird with setting the sample rate, though. At this point I'm just using the `with_sample_rate()` member function to boil down to `SupportedStreamConfig` and then the `config()` function to extract a `StreamConfig`.. Then I change the sample_rate member variable by hand. 

`SupportedStreamConfigRange::with_sample_rate()` seems to be malfunctioning though. It just returns the maximum supported sample_rate.. Which in some cases is 2**32-1.. Maybe I'll have to check again.. I'm running on Linux (ALSA).

**Comment by mitchmindtree** [comment]
> SupportedStreamConfigRange::with_sample_rate() seems to be malfunctioning though. It just returns the maximum supported sample_rate.. Which in some cases is 2**32-1.. Maybe I'll have to check again.. I'm running on Linux (ALSA).

@pietervandermeer I believe this bug was fixed in https://github.com/RustAudio/cpal/pull/462 though I don't recall whether or not this has been published yet. You could try switching your dependency to the master branch of this repo to see if this results in the specified sample rate.

**Comment by roderickvd** [comment]
Indeed, this has since been fixed.
